File: csw/mysql/jtest/src/GLPDB.java
package csw.mysql.jtest.src;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.ResultSet;

import java.io.*;

import java.util.NoSuchElementException;
import java.util.Scanner;
import java.util.ArrayList;



// The following comment is from a tutorial.
// I take it at face value and it seems to work.

// Notice, do not import com.mysql.jdbc.*
// or you will have problems!

public class GLPDB {

    String   g_dirname = null;
    double   g_xmin, g_ymin, g_xmax, g_ymax, g_xscal, g_yscal;
    long     g_ixmin, g_iymin, g_ixmax, g_iymax;
    Statement stmt = null;

// I have tried several values for the max number of rows
// per insert.  I think mysql can do more than 10000, but
// this java code seems to slow to a crawl with much more
// than 10000.

    final int MAX_FOR_INSERT = 10000;

    ResultSet  rs = null;

    public static Connection glpConnect () throws Throwable {

      try {

          // The following comment was on a tutorial site.  I
          // take it at face value and it seems to work this way,
          //
          // The newInstance() call is a work around for some
          // broken Java implementations

          Class.forName("com.mysql.jdbc.Driver").newInstance();

      }
      catch (Exception ex) {
          System.out.println ("Exception in LoadDriver main Method");
          throw (ex);
      }

      Connection  conn = null;

      try {
        conn =
           DriverManager.getConnection(
              "jdbc:mysql://localhost/glp?" +
              "user=glenn&password=Dog4-J3ssi3&useSSL=false");
      }
      catch (SQLException ex) {
        System.out.println ("Exception in getConnection stuff.");
        throw (ex);
      }

      return conn;
    }




    Connection  conn = null;

    GLPDB (Connection myconn) throws Throwable {
        if (myconn == null) {
            throw (new Exception ("Cannot use null connection."));
        }
        conn = myconn;
    }

//
// Fill the mysql spatial data values in the objects_geom table.
//
    public void fillSpIndex (int fid)
    {

      final String fdname = "/home/gpinkerton/data/load";

      //String fname = "csw/mysql/jtest/src/ld2.dat";
      String fname = fdname + "/load_objects_" + fid + ".dat";

      FileInputStream  fis = null;
      Scanner          fscan = null;

// create a scanner for the file

      try {
        fis = new FileInputStream (fname);
        fscan = new Scanner (fis);
      }
      catch (IOException ex) {
        System.out.println (ex.getMessage());
        return;
      }

      ArrayList<SpatialObj> alist = new ArrayList<SpatialObj> ();
      SpatialObj  sob = null;

      long   obid;
      int    itype;
      int    layer_id;
      double   xmin, ymin, xmax, ymax;

      FileWriter fw = null;

      try {
        fw = new FileWriter (fdname + "/jt_" + fid + ".sql");
      }
      catch (IOException ex) {
        System.out.println ("Cannot open jt.sql file");
        System.out.println (ex.getMessage());
        return;
      }

      int    n = 0;
      while (true) {
        try {
            obid = fscan.nextLong ();
            itype = fscan.nextInt ();
            layer_id = fscan.nextInt ();
            xmin = fscan.nextDouble ();          
            ymin = fscan.nextDouble ();          
            xmax = fscan.nextDouble ();          
            ymax = fscan.nextDouble ();          
            sob = new SpatialObj (obid, itype, layer_id,
                                  xmin, ymin, xmax, ymax);
            alist.add (sob);
            n++;
            if (n >= MAX_FOR_INSERT) {
              putIntoTable (alist, fw);
              alist.clear ();
              n = 0;
            }
        }

    // Should see this after end of file
        catch (NoSuchElementException ex) {
          putIntoTable (alist, fw);
          break;
        }
    // Any other exception is a problem
        catch (Throwable ex) {
          System.out.println ();
          System.out.println ("Problem exception scanning load input file.");
          System.out.println (ex.getMessage());
          System.out.println ();
          break;
        }
      }
      fscan.close();
      fscan = null;
      fis = null;

      try {
        fw.close ();
      }
      catch (IOException ex) {}

    }



    private void putIntoTable (ArrayList<SpatialObj> alist,
                               FileWriter fw)
    {

      if (alist.size() < 1) {
        return;
      }

      final String  s1 = "insert into objects_geom values \n";
      final String  b = " ";
      final String  r = ")";
      final String  l = "(";
      final String  c = ", ";
      final String  sst = "ST_PolygonFromText(";
      final String  ply = "'polygon((";
      final String  nl = "\n";

      int    ix1, iy1, ix2, iy2;
      long   id;
      SpatialObj  sob;

      String sq = new String (s1);
      String sln, sg, spt;
      
      int ndo = alist.size();

      for (int i=0; i<ndo; i++) {

        sob = alist.get (i);
        id = sob.object_id;
        ix1 = (int) (sob.xmin * 1000000.0);
        iy1 = (int) (sob.ymin * 1000000.0);
        ix2 = (int) (sob.xmax * 1000000.0);
        iy2 = (int) (sob.ymax * 1000000.0);

        spt = ply + ix1 + b + iy1 + c + nl +
                    ix1 + b + iy2 + c + nl +
                    ix2 + b + iy2 + c + nl +
                    ix2 + b + iy2 + c + nl +
                    ix1 + b + iy1 + b + r + r + "'" + nl;
        if (i < ndo - 1) {
          sg = sst + spt + r + r + c + "\n";
        }
        else {
          sg = sst + spt + r + r + "\n";
        }
        sln = l + id + c + sg;
        sq = sq + sln;

        if (i % 1000 == 0  &&  i > 0) {
          try {
            fw.write (sq);
            fw.flush ();
            sq = "";
          }
          catch (IOException ex) {
            System.out.println ("Error writing to jt.sql");
            System.out.println (ex.getMessage());
            break;
          }
        }
      }

      sq = sq + ";";

      try {
        fw.write (sq);
        fw.flush ();
      }
      catch (IOException ex) {
        System.out.println ("Error writing jt.sql");
        System.out.println (ex.getMessage());
      }

    }



    public void testMainStuff ()
    {
      doStuff ();
    }


    private void doStuff ()
    {

      String    sq, bbs;

// get global stuff
        
      sq = "select * from global_stuff";
      do_query (sq);

      getGlobalRsData (rs);
      try {
        rs.close ();
        rs = null;
        stmt.close ();
        stmt = null;
      }
      catch (SQLException exstmt) {
      }

      double x1, y1, x2, y2;
      Scanner scan = new Scanner( System.in );

      FileWriter  fw = null;

      try {
          fw = new FileWriter ("csw/mysql/jtest/src/j_bbox_sel.out");
      }
      catch (IOException ex) {
          System.out.println ();
          System.out.println ("Cannot open j_bbox_sel.out.");
          System.out.println (ex.getMessage());
          System.out.println ();
      }

long  stime, etime;
double  dtime;

      Scanner  lscan = null;
      String   sline = null;
 
      while (true) 
      {
       
        System.out.println (); 
        System.out.print ("Enter bbox: ");

        try {
          sline = scan.nextLine ();
          if (sline == null) break;
          if (sline.isEmpty()) break;
          lscan = new Scanner (sline);
          if (lscan.hasNextDouble() == false) break;
          x1 = lscan.nextDouble();
          y1 = lscan.nextDouble();
          x2 = lscan.nextDouble();
          y2 = lscan.nextDouble();
        }
        catch (Throwable ex) {
          break;
        }

stime = System.nanoTime ();

        bbs = createBboxString (x1, y1, x2, y2);

        sq = createSelectQuery (bbs);

//System.out.println ();
//System.out.println (sq);
//System.out.println ();

        do_query (sq);

        try {
          fw.write ("\n==== new search bbox =  " + sline + "  ====\n");
        }
        catch (IOException ex) {};

        getRsData (rs, fw);
        try {
          rs.close ();
          rs = null;
          stmt.close ();
          stmt = null;
        }
        catch (SQLException exstmt) {
        }

etime = System.nanoTime();

dtime = ((double)etime - (double)stime) / 1000000000.0;
System.out.println ("elapsed time = " + dtime);
System.out.println ();

      }
      
      scan.close();

      try {
        fw.close();
      }
      catch (IOException ex) { }

      return;

    }



    private void do_query (String sq)
    {

        try {
            stmt = conn.createStatement ();
            rs = stmt.executeQuery (sq);
        }
        catch (SQLException ex) {
            System.out.println ();
            System.out.println ("SQLException: " + ex.getMessage());
            System.out.println ("SQLState: " + ex.getSQLState());
            System.out.println ("VendorError: " + ex.getErrorCode());
            System.out.println ();
        }
    }



    private void getGlobalRsData (ResultSet rs)
    {

      try {
        rs.first ();

    // retrieve the single line from the global stuff table
        
        g_dirname = rs.getString ("dirname");
        g_xmin = rs.getDouble ("xmin");
        g_ymin = rs.getDouble ("ymin");
        g_xmax = rs.getDouble ("xmax");
        g_ymax = rs.getDouble ("ymax");
        g_ixmin = (long)rs.getInt ("ixmin");
        g_iymin = (long)rs.getInt ("iymin");
        g_ixmax = (long)rs.getInt ("ixmax");
        g_iymax = (long)rs.getInt ("iymax");
        g_xscal = (g_xmax - g_xmin) / (double)(g_ixmax - g_ixmin);
        g_yscal = (g_ymax - g_ymin) / (double)(g_iymax - g_iymin);

      }
      catch (SQLException ex) { }

    }



/**
 * Loop through the result set returned for the "spatial"
 * select operation.  The result set is orderedd by file id
 * and then file pos.  Thus, when a new file id is encountered
 * a new random access file object is created. Since the set
 * is ordered primarily by file id, this should not produce
 * an excess of file opens and file closes.
 */
    private void getRsData (ResultSet rs,
                            FileWriter fw)
    {

      long       obid = 0;
      int        line_id, file_id, file_pos;
      RandomAccessFile fp = null;

      int        last_file_id = -1;

      try {
        rs.last();
        int nr = rs.getRow();
        try {
          fw.write ("     number of rows found = " + nr + "\n");
        }
        catch (IOException ex) {};
        rs.beforeFirst ();
        while (rs.next ()) {
          obid = rs.getLong ("object_id");
          line_id = rs.getInt ("line_id");
          file_id = rs.getInt ("file_id");

          if (file_id != last_file_id) {
            try {
              if (fp != null) {
                fp.close();
                fp = null;
              }
              fp = new RandomAccessFile (g_dirname + "/glp_points_" +
                                         file_id + ".dat", "r");
              last_file_id = file_id;
            }
            catch (IOException ex) {
              System.out.println ();
              System.out.println ("Cannot open random access points file.");
              System.out.println (ex.getMessage());
              System.out.println ();
              break;
            }
          }

          file_pos = rs.getInt ("file_pos");
          processObj (fw, fp,
                      obid, line_id, file_id, file_pos);
        }
      }
      catch (SQLException ex) {
        System.out.println ();
        System.out.println ("SQLException from getRsData");
        System.out.println (ex.getMessage());
        System.out.println ();
      }

      if (fp != null) {
        try {
          fp.close();
        }
        catch (IOException ex) {}
        fp = null;
      }

      try {
        int count = 0;
        rs.last();
        count = rs.getRow();
        rs.beforeFirst();
System.out.println ();
System.out.println ("Number of results = " + count);
System.out.println ();
      }
      catch (SQLException ex) { }

    }


    private void processObj (FileWriter fw,
                             RandomAccessFile fp,
                             long obid,
                             int  line_id,
                             int  file_id,
                             int  file_pos)
    {

      try {
        fw.write ("\nobject_id = " + obid + "\n");
        fw.write ("file_id = " + file_id + "\n");
        fw.write ("file_pos = " + file_pos + "\n");
        outputPointsFromFile (fw, fp, file_pos);
      }
      catch (IOException ex) {
        System.out.println ("Error writing results to file.");
        System.out.println (ex.getMessage());
      }
      
    }


    private void outputPointsFromFile (FileWriter fw,
                               RandomAccessFile fp,
                               int file_pos)
    {
      int   idum = 0;
      int   npts = 0;
      int   nbytes = 0;
      int   ix = 0;
      int   iy = 0;
      UniversePoint  up = null;
      try {
        fp.seek ((long)file_pos);
        nbytes = fp.readInt ();
        idum = fp.readInt ();
        npts = fp.readInt ();
        fw.write ("nbytes = " + nbytes + "  npts = " + npts + "\n");
        for (int j=0; j<npts; j++) {
          ix = fp.readInt ();
          iy = fp.readInt ();
          up = convertIntToUniverse (ix, iy);
          fw.write ("  x = " + String.format("%.5f", up.x) + 
                      "  y = " + String.format("%.5f", up.y) + "\n");
        }
        fw.flush();
        
      }
      catch (IOException ex) {
        System.out.println ("Error reading point file data");
      }
        
    }
         

    private UniversePoint convertIntToUniverse (int ix, int iy) 
    {
      double  dx, dy;  
      dx = (double)((long)ix - g_ixmin) * g_xscal + g_xmin;
      dy = (double)((long)iy - g_iymin) * g_yscal + g_ymin;
      UniversePoint up = new UniversePoint (dx, dy);
      return up;
    }


    private class UniversePoint {
      double   x = 0.0,  y = 0.0;
      UniversePoint (double xin, double yin) {
        x = xin;
        y = yin;
      }
    }




/**
 * Build a string of sql that will return objects inside the
 * specified bounding box.
 */
    private String  createBboxString (double    bxmin, 
                                      double    bymin,
                                      double    bxmax,
                                      double    bymax)
    {
      String  s;

      int  ix1 = (int)(bxmin * 1000000.0);
      int  iy1 = (int)(bymin * 1000000.0);
      int  ix2 = (int)(bxmax * 1000000.0);
      int  iy2 = (int)(bymax * 1000000.0);

      s = 
        "'polygon ((\n" +
           ix1 + " " + iy1 + " , \n" +
           ix1 + " " + iy2 + " , \n" +
           ix2 + " " + iy2 + " , \n" +
           ix2 + " " + iy1 + " , \n" +
           ix1 + " " + iy1 + " \n" + "))'";

      return s;   
    }

/**
 * Build a string of sql that will return objects inside the
 * specified bounding box.
 */
    private String  createSelectQuery (String bbs)
    {
      String  s;

      s = "select objects.object_id,pline_points_lookup.* from \n" +
          "objects_geom,objects,pline_id_lookup,pline_points_lookup \n" +
          " where \n" +
          "(MBRIntersects" + "\n" +
             "(ST_GeomFromText\n(" +
               bbs + "),\n objects_geom.bbox)) \n" +
          "  and \n" +
          "(objects.object_id = objects_geom.object_id) \n" +
          "  and \n" +
          "(pline_id_lookup.object_id = objects.object_id) \n" +
          "  and \n" +
          "(pline_points_lookup.line_id = pline_id_lookup.line_id) \n" +
          "order by pline_points_lookup.file_id, pline_points_lookup.file_pos"
      ;

      return s;   
    }


} // end of GLPDB class

File: csw/python/mysql/glp_db.py

#  This is a class to access the glp mysql data base.  The 
#  purpose of this is to learn a bit more python in a somewhat
#  realistic problem.


#  Most of the web instructions for using mysql (at least those
#  which I looked at) say to import PyMySQL.  The PyMySQL does
#  not work on my linux box (fedora 26).  I found a suggestion
#  to use all lower case, and that works for me.

import pymysql

# regular expression stuff and timing stuff
import re
import time


#  The glp_fileio module has the GLPFileIO class, which reads
#  data from the random access points files associated with the
#  glp database.

import glp_fileio


# class to hold id values (object, file, file pos) that are
# returned by a "spatial" query of the database

class ObjFilePos:
  def __init__ (self, obid, f_id, f_pos):
    self.object_id = obid
    self.file_id = f_id
    self.file_pos = f_pos




#  Encapsulate the glp database stuff in class GLPDB
#  This is a very simple "spatial" data base which
#  can be populated with random "spatial" objects
#  so I (Glenn) can work with mysql in python.

class GLPDB:

#  The constructor needs to establish connection to the mysql dbase.

  def __init__ (self):

    self.dirname = ""
    self.connection = pymysql.connect (host = "localhost",\
                                       user = "glenn",\
                                       password = "Dog4-J3ssi3",\
                                       db = "glp")
    try:
      with self.connection.cursor() as curs:
        sq = "select * from global_stuff"
        curs.execute(sq)
        for (dirname, xmin, ymin, xmax, ymax, ixmin, iymin, ixmax, iymax) in curs:
          self.dirname = dirname
          self.xmin = xmin
          self.ymin = ymin
          self.xmax = xmax
          self.ymax = ymax
          self.ixmin = ixmin
          self.iymin = iymin
          self.ixmax = ixmax
          self.iymax = iymax
          self.fio = glp_fileio.GLPFileIO (xmin, ymin, xmax, ymax,\
                                ixmin, iymin, ixmax, iymax)
          break

    finally:
      pass

# !! end of __init__ constructor


# Method to execute the bbox select query

  def __do_bbox_query_ (self, sq):

    rlst = []

    with self.connection.cursor() as curs:
      curs.execute(sq)
      for (object_id, line_id, file_id, file_pos) in curs:
        ofp = ObjFilePos (int(object_id), int(file_id), int(file_pos))
        rlst.append(ofp)

      return rlst          

# !! end of __do_bbox_query_ method
    
    
#  Method to disconnect from the mysql server

  def disconnect (self):
    self.connection.close()

# !! end of disconnect method


# create an sql string to select objects intersecting
# the specified rectangle

  def __create_bbox_query_ (self, bxmin, bymin, bxmax, bymax):

    ix1 = int(bxmin * 1000000.0)
    iy1 = int(bymin * 1000000.0)
    ix2 = int(bxmax * 1000000.0)
    iy2 = int(bymax * 1000000.0)

    bbs =\
      "'polygon ((\n" +\
         str(ix1) + " " + str(iy1) + " , \n" +\
         str(ix1) + " " + str(iy2) + " , \n" +\
         str(ix2) + " " + str(iy2) + " , \n" +\
         str(ix2) + " " + str(iy1) + " , \n" +\
         str(ix1) + " " + str(iy1) + " \n" + "))'"

    s = "select objects.object_id,pline_points_lookup.* from \n" +\
          "objects_geom,objects,pline_id_lookup,pline_points_lookup \n" +\
          " where \n" +\
          "(MBRIntersects" + "\n" +\
             "(ST_GeomFromText\n(" +\
               bbs + "),\n objects_geom.bbox)) \n" +\
          "  and \n" +\
          "(objects.object_id = objects_geom.object_id) \n" +\
          "  and \n" +\
          "(pline_id_lookup.object_id = objects.object_id) \n" +\
          "  and \n" +\
          "(pline_points_lookup.line_id = pline_id_lookup.line_id)\n" +\
          "order by pline_points_lookup.file_id, pline_points_lookup.file_pos"

    return s

# !! end of __create_bbox_query_ method



#  function to prompt for bbox limits, get the obkects intersecting
#  the bbox and wite them to a file.

  def test_bbox (self):

    BBERR = "Illegal bbox definition, try again"
    while True:

      try:
        print ("")
        c = input("Enter lower left x y and upper right x y: ")
        if len(c) < 1:
          break

        try:
          bblst = re.split(" ", c)
        except Exception as e:
          print ("Exception from re.split: " + str(e))
          continue

        rpair = self.get_bbox_objects (float(bblst[0]), float(bblst[1]),
                                       float(bblst[2]), float(bblst[3]))

      except Exception as e:
        print ("Exception in test_bbox: " + str(e))
        raise

      sel_count = rpair[0]
      delta_time = rpair[1]

      print ("")
      print ("")
      print ("number of objects selected = " + str(sel_count))
      print ("")
      print ("elapsed time for query = " + str(delta_time))
      print ("")

  #  ! end of while true loop to prompt for bbox limits and process them


  # comes here if no exception happened 
  # still may need to close the output and point files

    try:
      if not fout.closed:
        fout.close()
      if not fpts.closed:
        fpts.close()
    except:
      pass

    return
      

# !! end of test_bbox method




#  function to prompt for bbox limits, get the obkects intersecting
#  the bbox and wite them to a file.

  def get_bbox_objects (self, xmin, ymin, xmax, ymax):

    try:
      fout = open ("py_bbox_sel.out", "w")
    except IOError:
      print ("****** Cannot open file: py_bbox_sel.out ******")
      return

    return_pair = []

    BBERR = "Illegal bbox definition, try again"
    try:

        tstart = time.time()

        sq = self.__create_bbox_query_ (xmin, ymin, xmax, ymax)
        file_pos_list = self.__do_bbox_query_ (sq)
        sel_count = len(file_pos_list)
        sl = str(sel_count)

        return_pair.append(sel_count)

        c = str(xmin) + " " + str(ymin) + " " + str(xmax) + " " + str(ymax)
        fout.write ("\n==== new search bbox =  " + c + "  ====\n")
        fout.write ("     number of rows found = " + sl + "\n")
  #      print ("\n     number of rows found = " + sl + "\n")

        last_file_id = -1
        for ofp in file_pos_list:
          fout.write ("\n")
          fout.write ("object_id = " + str(ofp.object_id))
          fout.write ("\n")
          fout.write ("file_id = " + str(ofp.file_id))
          fout.write ("\n")
          fout.write ("file_pos = " + str(ofp.file_pos))
          fout.write ("\n")

          if last_file_id != ofp.file_id:
            try:
              if last_file_id != -1  and  not fpts.closed:
                fpts.close()

              fpts = open (self.dirname + "/glp_points_" +\
                           str(ofp.file_id) + ".dat", "rb")
              last_file_id = ofp.file_id
            except IOError:
              print ("")
              print ("Error opening points random access file " +\
                      str(ofp.file_id))
              print ("")
              raise

       # end of open file_id block

          plist = self.fio.read_points (fpts, ofp.file_pos)
          fout.write (" npts = " + str(len(plist)))
          fout.write ("\n")
          for upt in plist:
       # the f just prior to the quote signals a "formatted" string
       # the curly braces surround value:format pairs.
            fout.write(f"  x = {upt.x:.5f}  y = {upt.y:.5f}")
            fout.write ("\n")


    except Exception as e:
        print ("Exception in test_bbox: " + str(e))
    # close output file and point before raising the exception
        if not fout.closed:
          fout.close()
        if not fpts.closed:
          fpts.close()
        raise

    tend = time.time() 
    delta_time = round(tend - tstart, 2)

    return_pair.append(delta_time)

  # comes here if no exception happened 
  # still probably need to close the output and point files
    if not fout.closed:
      fout.close()

    try:
      if not fpts.closed:
        fpts.close()

    except:
      pass

    return return_pair
      

# !! end of get_bbox_objects method



## !!! end of GLPDB class definition




# make a class and print the first "line" points in file
# This code is not part of the GLPDB class

if __name__ == "__main__":
  db = GLPDB ()
  db.test_bbox()



